### ISSUE TITLE
Data parallel problem (with multi GPUs)

### ISSUE THREAD

**Issue by bcd33bcd** [caogang/wgan-gp#22]
I try to implement your code with multi-gpus, but the network occur the problem with 

> "RuntimeError : Scatter is not differentiable twice"

 while add "torch.dataparallel()" to the Discriminator and Generator.
Followed by the suggestion of [""](https://github.com/jfsantos/dragan-pytorch/blob/master/dragan.py)， the error changes to 

> "RuntimeError : arguments are located on different GPUs at /opt/conda/conda-bld/pytorch_1503966894950/work/torch/lib/THC/generated/../generic/THCTensorMathPointwise.cu:215"

How could I fix the problem? Thank you.

**Comment by caogang** [comment]
Maybe you should make the arguments located on a same GPU?. Could you show the code?

**Comment by bcd33bcd** [comment]
Thank you for your reply.
Here is the model of discriminator 
`class Discriminator(nn.Module):
    def __init__(self,nc,ndf):
        super(Discriminator,self).__init__()
        # 32 x 32
        self.layer1 = nn.Sequential(nn.Conv2d(nc,ndf,kernel_size=4,stride=2,padding=1),
                                 nn.BatchNorm2d(ndf),
                                 nn.LeakyReLU(0.2,inplace=True))
        self.layer1 = nn.DataParallel(self.layer1)
        # 16 x 16
        self.layer2 = nn.Sequential(nn.Conv2d(ndf,ndf*2,kernel_size=4,stride=2,padding=1),
                                 nn.BatchNorm2d(ndf*2),
                                 nn.LeakyReLU(0.2,inplace=True))
        self.layer2 = nn.DataParallel(self.layer2)
        # 8 x 8
        self.layer3 = nn.Sequential(nn.Conv2d(ndf*2,ndf*4,kernel_size=4,stride=2,padding=1),
                                 nn.BatchNorm2d(ndf*4),
                                 nn.LeakyReLU(0.2,inplace=True))
        self.layer3 = nn.DataParallel(self.layer3)
        # 4 x 4
        self.layer4 = nn.Sequential(nn.Conv2d(ndf*4,2,kernel_size=4,stride=1,padding=0),
                                 nn.Sigmoid())
        self.layer4 = nn.DataParallel(self.layer4)
        

    def forward(self,x):
        out = self.layer1(x)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        return out`


And in the main.py, there is 
`
netD = torch.nn.DataParallel(Discriminator(nc, ndf))
netG = torch.nn.DataParallel(Generator(nc, ngf, opt.nz))
if(opt.cuda):
    netD.cuda()
    netG.cuda()

` 

Variables is initilized by 
`##########   GLOBAL VARIABLES   ###########
noise = torch.FloatTensor(opt.batchSize, opt.nz, 1, 1)
real = torch.FloatTensor(opt.batchSize, nc, opt.imageSize, opt.imageSize)
# label = torch.FloatTensor(opt.batchSize)
real_label = 1
fake_label = 0
real_list = torch.FloatTensor(opt.batchSize)
fake_list = torch.FloatTensor(opt.batchSize)

noise = Variable(noise)
real = Variable(real)
# label = Variable(label)
real_list = Variable(real_list)
fake_list = Variable(fake_list)

if(opt.cuda):
    noise = noise.cuda()
    real = real.cuda()
    # label = label.cuda()
    real_list = real_list.cuda()
    fake_list = fake_list.cuda()

`

Training:
`        # train with real data, resize real because last batch may has less than
        # opt.batchSize images
        real.data.resize_(images.size()).copy_(images)
        # label = label.data.resize_(images.size(0),2)
        fake_list = fake_list.data.resize_(images.size(0),1).fill_(fake_label)
        real_list = real_list.data.resize_(images.size(0),1).fill_(real_label)
        label = Variable(torch.cat((fake_list,real_list),1))
        if opt.cuda:
        	label.cuda()


        errD_real = netD(real)
        prob = nn.functional.softmax(errD_real)
        # errD_real = torch.mean(errD_real, dim=1)
        errD_real = criterion(prob, label)
        errD_real.backward()`

Please make the point of my mistake. Thanks.

**Comment by jgbos** [comment]
This error does not appear to be related to multiple gpus itself, but rather the use of using the `nn.parallel` functions.  If I use the following:

```python
def forward(self, input):
        return nn.parallel.data_parallel(self.forward_, input, self.gpu_ids)
```

I get the error even though I'm only using one GPU, but if I just do

```python
def forward(self, input):
        return self.forward_(input)
```

I do not get any errors (input and the model are loaded on the current GPU).
